code/software/tests/malloc-test/kmain.c

```diff
@@ -20,24 +20,141 @@
 #include <stdlib.h>
 #include <string.h>
 #include <heap.h>
+#include <machine.h>
 
 static const char *test = "TESTING!TESTING!TESTING!TESTING!";
 
 extern int newB();
 extern int getBNum();
 extern int deleteB();
 
+void calloc_check(size_t count, size_t size) {
+    printf("calloc zero check      : starting\n");
+    size_t total = count * size;
+    bool success = true;
+
+    for (unsigned loop = 0; loop < 10 && success; ++loop) {
+        unsigned char *check_tmp = calloc(count, size);
+        if (!check_tmp) {
+            printf("calloc zero check      : failed to calloc(0x%zX, 0x%zX)\n", count, size);
+            success = false;
+            break;
+        }
+
+        for (size_t i = 0; i < total; ++i) {
+            if (check_tmp[i] != 0) {
+                printf("calloc zero check      : failed at loop %u byte 0x%zX\n", loop, i);
+                success = false;
+                break;
+            }
+            check_tmp[i] = i ^ loop;  // Fill with garbage
+        }
+        free(check_tmp);
+    }
+
+    if (success) {
+        printf("calloc zero check      : succeeded\n");
+    }
+}
+
+void calloc_benchmark(size_t count, size_t size) {
+    printf("calloc benchmark       : starting\n");
+    bool success = true;
+    unsigned start_ticks = _TIMER_100HZ;
```

Comment:
Hmm, unless I'm missing something, I think you'd want this to be something like:

```
unsigned start_ticks = *(uint32_t*)&_TIMER_100HZ;
```

(I'll admit these linker `PROVIDE`s always cause me some thinking, but something like the above is what I've generally used in other code and found to work 🤔 )

Reply:
Hmm, I'll have to test that out once I've got the system tick fix up and running. Though looking at `machine.h`, it looks like `_TIMER_100HZ` is already `extern volatile uint32_t`, so I'm not sure that'll change things, aside from casting away the `volatile`.

Reply:
Hmm, yes, you're right. It should _just work_ . I just did a little test here, and it seems fine:

```
TEST(system_tick_is_running) {
    uint32_t start_ticks = _TIMER_100HZ;

    do {
        opt_guard++;
    } while (opt_guard < 1000000);

    uint32_t tt= _TIMER_100HZ;
    printf("_TIMER_100HZ = 0x%08lx\n", tt);

    ASSERT(start_ticks != _TIMER_100HZ);

    return true;
}
```

Gives:

```
_TIMER_100HZ = 0x00067266
 * system_tick_is_running - _TIMER_100HZ = 0x000673c4
PASSED
```

So please ignore my previous comment :D